sktime_dl/deeplearning/lstm/_regressor.py

```diff
@@ -0,0 +1,95 @@
+__author__ = "Withington"
+
+from tensorflow import keras
+
+from sktime_dl.deeplearning.base.estimators import BaseDeepRegressor
+from sktime_dl.deeplearning.lstm._base import LSTMNetwork
+from sktime_dl.utils import check_and_clean_data
+
+
+class LSTMRegressor(BaseDeepRegressor, LSTMNetwork):
+    ''' Long Short-Term Memory (LSTM)
+
+    Adapted from the implementation of Brownlee, J. (2018)
+
+    https://machinelearningmastery.com/how-to-develop-lstm-models-for-time-series-forecasting/
+    '''
+
+    def __init__(self,
+                 nb_epochs=200,
+                 batch_size=16,
+                 units=[50, 50],
+                 random_seed=0,
+                 verbose=False,
+                 model_name="lstm_regressor",
+                 model_save_directory=None):
+        super().__init__(
+            model_name=model_name,
+            model_save_directory=model_save_directory)
+        LSTMNetwork.__init__(
```

Comment:
We've been doing this in other places as well, but this way of handling multiple inheritance doesn't seem right to me. 

`model_name` and `model_save_directory` should be part of the `BaseDeepNetwork` as we use them in both classifiers and regressors, so `LSTMNetwork` should inherit from `BaseDeepNetwork` and call its constructor via `super()`. This works without any problems for the `BaseDeepRegressor`, which essentially becomes a mixin class in that case, but the problem is that `BaseDeepClassifier` creates other attributes (`classes` and `nb_classes`), so I believe we have three options:
* leave it as it is,
* make classes [cooperative](https://rhettinger.wordpress.com/2011/05/26/super-considered-super/),
* simply don't initiate the attributes in the constructor and create them when needed as done in sklearn 

Also see this [stackoverflow post](https://stackoverflow.com/questions/39415085/python-calling-init-for-multiple-parent-classes) and [linked article](https://fuhm.net/super-harmful/).

Reply:
Finer details of python inheritance, and what looks good to a mainly python developer vs what functionally works etc. are probably best for me to delegate. 

An alternative overall is potentially also to use composition for the networks? 

Currently, LSTMRegressor _is a_  BaseDeepRegressor, and _is a_ LSTMNetwork

Could reorganise so that LSTMRegressor _is (still) a_  BaseDeepRegressor, and _has a_ LSTMNetwork

It makes no real difference at all for this class (aside from aesthetics), but for those classifiers with bespoke training methods and/or internal parameter tuning, e.g. MCNN, maybe it makes sense to say that it _is_ a regressor that _has_ a network + a method of training/tuning it (which, for most architectures, is just to call the data cleaning functions and the basic keras fit()). 

Ultimately though, this code is 'supposed to be' hidden from the user, aside from being open source obviously. Some janky assumptions and dependencies behind the scenes are permissible in my opinion, assuming this doesn't become a functional as well as aesthetic issue in the future

Reply:
Interesting, thanks.
- Cooperative classes might work, yes. It would be nice to see how it looks here. Do you have time to try it out @mloning ?
- Initiate attributes when needed - yes, straightforward. Downsides: 
   - either a lot of code repetition (but, since 95% of it is assignment, it's only as much as calling `super.__init__` in many places) and an increased risk of missing initiating one of the attributes somewhere
   - or
   - writing an `initiate_common_variables` function.
- Regressor _has a_ network does seem appropriate and was [trialed](https://github.com/Withington/sktime-dl/blob/4ba698eb4a377d52ff76e9859caae59a0076d236/sktime_dl/regressors/deeplearning/_cnn.py) but hit problems when forecasting called into [sklearn clone](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/base.py). The `BaseDeepNetwork` parameters were not cloned.

Reply:
* So with cooperative classes it would look something like the code below. Technically, B's call to super isn't required, but this way the order of inheritance doesn't matter. 
 
```python
class A:  # base network
    
    def __init__(self, a, **kwargs):
        self.a = a
        super(A, self).__init__(**kwargs)

class B:  # base regressor/classifier
    
    def __init__(self, b, **kwargs):
        self.b = b
        super(B, self).__init__(**kwargs)

class C(A, B):  # concrete estimator
    
    def __init__(self, a, b, c):
        self.c = c
        super(C, self).__init__(a=a, b=b)

c = C(1, 2, 3)
print(c.__dict__)
>>> {'c': 3, 'a': 1, 'b': 2}
```
* But I think I prefer to avoid this type of multiple inheritance here, and create attributes on the fly, we can add unit tests that check that all classifiers have the required attributes after calling fit. 
* I didn't understand why clone didn't work? I think we should ensure that clone works on all our estimators, that's essential for tuning and ensembling.

Reply:
Our estimators work fine with clone now. I put a link above to an old commit where clone failed; it was when using BaseDeepRegressor _has a_ BaseDeepNetwork.

Have I read it right: you favour this option below?

> simply don't initiate the attributes in the constructor and create them when needed as done in sklearn

Reply:
Yes, but don't have a strong preference, I'd still prefer cooperative classes over the current way I think.

Reply:
I opened up issue #45 for this because it will need a bit of work, separate to LSTM.